### PR TITLE
Add `ListPackages()` to persistence service

### PR DESCRIPTION
**Pull request by djjuhasz**
Refs #988

This continues the work toward #988 by adding a `ListPackages()` method to the persistence service that includes a count of the total search results returned (before paging), as well as allowing filtering, sorting, and paging of the search.

Other changes:
- Add a telemetry wrapper for the `ListPackages()` method to provide extra performance and debugging data
- Add additional commentary for the `Filter.Page()` method
- Add many unit tests for the `ListPackages()` functionality

### CODE REVIEW

File: internal/persistence/ent/client/package.go [hunk 1]
```diff
@@ -130,3 +132,56 @@ func (c *client) UpdatePackage(
...
+		Offset: f.Offset,
+		Total:  total,
+	}
+
+	return res, pp, err
+}
+
+// filterPackages filters a package query based on filtering inputs.
+func filterPackages(q *db.PkgQuery, f *persistence.PackageFilter) (page, whole *db.PkgQuery) {
+	h := NewFilter(q, SortableFields{
```
Comment (jraddaoui) [inline internal/persistence/ent/client/package.go]: Why `h`?

File: internal/persistence/ent/client/package_test.go [hunk 2]
```diff
@@ -10,14 +10,26 @@ import (
...
 
+func nullUUIDToPtr(u uuid.NullUUID) *uuid.UUID {
+	var p *uuid.UUID
+	if u.Valid {
+		p = &u.UUID
+	}
+
+	return p
+}
+
```
Comment (jraddaoui) [inline internal/persistence/ent/client/package_test.go]: Only used twice with UUIDs already known to be valid. I's use `&u.UUID` directly in those two cases.

**Review by jraddaoui — APPROVED** [review]
LGTM, thanks @djjuhasz!

File: internal/persistence/ent/client/package.go (hunk 1 above)
Comment (djjuhasz) [reply internal/persistence/ent/client/package.go]: I'm not sure — I copied it over from the STE implementation as-is.

File: internal/persistence/ent/client/package_test.go (hunk 2 above)
Comment (djjuhasz) [reply internal/persistence/ent/client/package_test.go]: Removed.

File: internal/persistence/ent/client/package.go (hunk 1 above)
Comment (djjuhasz) [reply internal/persistence/ent/client/package.go]: I'm open to other suggestions.

Comment (jraddaoui) [reply internal/persistence/ent/client/package.go]: Maybe changing the parameter var name too? 

```
func filterPackages(q *db.PkgQuery, pf *persistence.PackageFilter) (page, whole *db.PkgQuery) {
	f := NewFilter(q, SortableFields{...
```

Comment (djjuhasz) [reply internal/persistence/ent/client/package.go]: I tried changing the `f *persistence.PackageFilter` param to `pf *persistence.PackageFilter` but I didn't like that the new name didn't agree with the `ListPackages(ctx context.Context, f persistence.PackageFilter)` name — I think passing through PackageFilter with the same variable name is easier to track.

I've gone with changing the query filter name from `h` to `qf` — it's a bit longer but I kind of like that the parameters `q` and `f` become query filter `qf`. Math nerd joy :nerd_face:

**Pull request merged**